File: server/common/socket.py
from common.helpers import close_socket

class Socket:
    def __init__(self, socket):
        self._skt = socket
    
    def receive_all(self, amount_to_receive):
        '''
        Receives a specific amount of bytes from the socket, ensuring no short reads
        If the connection is closed before receiving all data, 
        raises a ConnectionError.
        '''
        try:
            received_data = b''
            while len(received_data) < amount_to_receive:
                chunk = self._skt.recv(amount_to_receive - len(received_data))
                if not chunk and not received_data:
                    return None
                if not chunk:
                    raise ConnectionError("Socket connection closed before receiving all data")    
                received_data += chunk
            return received_data
        except OSError:
            raise ConnectionError("Socket was manually closed during receive")
    
    def send_all(self, data):
        """
        sends all the data to the socket, ensuring no short writes
        """
        try:
            self._skt.sendall(data)
        except OSError:
            raise ConnectionError("Failed to send all data")

    def close(self):
        """
        Closes the socket connection
        """
        close_socket(self._skt, "client")
        

File: client/common/socket.go
package common

import (
	"math"
	"net"
	"time"
)

// Socket is a wrapper around net.Conn that provides methods for sending and receiving data,
// ensuring that all data is sent or received as needed.
type Socket struct {
	conn net.Conn
}

// NewSocket creates a new Socket instance by connecting to the specified server address.
// It returns a pointer to the Socket and any error encountered during the connection process.
func NewSocket(serverAddress string) (*Socket, error) {
	maxAttemps := 3
	var err error
	for i := 0; i < maxAttemps; i++ {
		conn, err_aux := net.Dial("tcp", serverAddress)
		if err_aux == nil {
			return &Socket{conn}, nil
		}
		err = err_aux

		// Exponential backoff strategy for reconnection attempts
		waitTime := time.Duration(math.Pow(2, float64(i))) * time.Second
		time.Sleep(waitTime)
	}
	return nil, err
}

// SendAll sends the entire byte slice data to the connected server.
// It ensures that all bytes are sent.
// If an error occurs during sending, it returns the error.
func (s *Socket) SendAll(data []byte) error {
	totalSent := 0
	for totalSent < len(data) {
		sended, err := s.conn.Write(data[totalSent:])
		if err != nil {
			return err
		}
		totalSent += sended
	}
	return nil
}

// ReceiveAll receives a specified amount of bytes from the connected server.
// It ensures that the exact amount of bytes is received and returns the data as a byte slice.
// If an error occurs during receiving, it returns the error.
func (s *Socket) ReceiveAll(amountToReceive int) ([]byte, error) {
	buffer := make([]byte, amountToReceive)
	total_received := 0
	for total_received < amountToReceive {
		received, err := s.conn.Read(buffer[total_received:])
		if err != nil {
			return nil, err
		}
		total_received += received
	}
	return buffer, nil
}

// Close closes the connection to the server.
// It returns any error encountered during the closing process.
func (s *Socket) Close() error {
	return s.conn.Close()
}
